### ISSUE TITLE
Timeout don't work while using SyncCommand

### ISSUE THREAD

**Issue by RongieZeng** [hudl/Mjolnir#54]
Hi,
Netflix's Hystrix component is very useful, so i'm looking for the same component in C#, and i search the google and find your project. But it seems the timeout future don't work though i've pass it to the syncCommand's constructor.  This is the very import future i think, please let me know is it the Mjolnir can't do it with sync invoke, or i use it in a wrong way?

**Comment by robhruska** [comment]
Could you describe a bit more about what's not working with it? Is your Command a network call? If you could share a code snippet (maybe your Command's execute method contents), that'd be helpful.

The timeout should work for any network calls within the Command's execute method, provided that you use the passed-in `CancellationToken` and pass it through to your own calls.

Mjolnir currently doesn't do any thread aborts (i.e. it won't hard-kill a thread or operation unless that operation's also looking at the `CancellationToken` during its execution).

[This page](https://msdn.microsoft.com/en-us/library/dd997364(v=vs.110).aspx) might help shed some insight if you haven't run across it already.

**Comment by RongieZeng** [comment]
It's a wcf service calling, so there is no CancellationToken parameter to pass into. 
`public class GetUserCommand : SyncCommand<UserResp>
{
    private UserReq _userReq;
    public GetUserCommand(UserReq userReq)
        : base("Vipcenter", "UserInfo", TimeSpan.FromSeconds(1))
    {
        _userReq = userReq;
    }

    public override bool Execute(CancellationToken cancellationToken)
    {
        using(var userServiceClient = new UserServiceClient())
        {           
            var userResp = userServiceClient.GetUser(_userReq);
            return userResp;
        }
   }
}`

Inside the GetUser Service, i make a Thread.Sleep(3000) there, but although i set the timeout with 1 second, the excute method don't fail fast

**Comment by robhruska** [comment]
It's unfortunate that the service method there doesn't have a CancellationToken you can pass through - I'm not sure I have a great suggestion.

All of our cancellation/timeouts are cooperative, so we've never had a strong need to address the case where an API/library doesn't support timeouts.

Is `UserServiceClient` something within your control that you can add `CancellationToken` support to?

**Comment by RongieZeng** [comment]
UserServiceClient is a subclass of ClientBase<T>, which is generated by svcutil.exe of visual studio. Do you mean that i can wrap my GetUser calling in another method, like this:
`
    public Task<UserResp> GetUserAsyc(CancellationToken cancellationToken)
    {
            await new TaskFactory().StartNew<UserResp>(()=>{
                 using(var userServiceClient = new UserServiceClient())
                {           
                    var userResp = userServiceClient.GetUser(_userReq);
                    return userResp;
                }
            }, cancellationToken);        
    }

`
And then:
`
        public override UserResp Execute(CancellationToken cancellationToken)
        {
            return GetUserAsyc(cancellationToken)
        }
`

**Comment by robhruska** [comment]
I'd be wary of doing that - it might "work", but you're still going to have the `GetUser` call hanging out there after the task cancellation happens.

You might find these useful for some more context:
- http://www.ben-morris.com/why-you-shouldnt-create-asynchronous-wrappers-with-task-run/
- http://blog.stephencleary.com/2013/08/startnew-is-dangerous.html

**Comment by RongieZeng** [comment]
Thank you for your suggestion. But the UserServiceClient.GetUser is a RPC calling,  it's a remoting service in another system, i can't control it

**Comment by robhruska** [comment]
Any sort of network I/O API that doesn't allow for passing or configuring a timeout is putting you in kind of a tight spot, unfortunately. What kind of service call is it? Is it WCF?

**Comment by RongieZeng** [comment]
Yes, it's wcf, and i know wcf support timeout itself. But in my system, there is other services that don't support timeout, and this is the important reason why i want to find a component like hystrix

**Comment by robhruska** [comment]
Gotcha. Unfortunately, killing threads/operations isn't something that Mjolnir supports at the moment. It's a possibility down the road, and I'd be happy to help advise on a PR if you'd be interested in contributing!
